Migration de la CI de travis à Github Actions

firm1 opened the pull request:
On en parlait il y a quelques jours de migrer sur github actions ,après quelques essais fructueux voici une proposition qui est maintenant à un stade avancé.

### Qu'est-ce qui a été fait ?

La migration de travis vers github actions a permise de revoir un peu l'architecture des jobs.

Voici comment les jobs sont retranscris sous github actions.

![mermaid-diagram-20201122154811](https://user-images.githubusercontent.com/6066015/99908927-bf16af80-2ce5-11eb-829f-01934ea88c77.png)

On passe d'un temps d’exécution de **13min** (sous travis) à **11min** (sous github actions).

Certaines choses ont été simplifiées (avant on avait un peu trop de condition de déclenchement des installs) et les scripts liés à travis sont supprimés de la PR.

### Ce qu'il reste a faire

- Mettre à jour dans le workflow avec les bonnes variables :
    - `BOT_LOGIN`
    - `BOT_PASSWORD`
- Tester le push de la doc sur github

### Contrôle qualité

Il sera fait depuis Github actions

- Allez dans les [actions du projet zds-site](https://github.com/zestedesavoir/zds-site/actions) et vérifiez que le jobs s’exécute avec succès

firm1 committed:
remove .travis.yml

AmauryCarrade commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,209 @@
+name: Tests zds-site
```
On va croiser ce nom partout, et il prend de la place quand on a des tâches qu'ont de biens longs noms. De plus il n'est pas représentatif car en l'état ce _workflow_ ne fait pas que tester.

Ok, là c'est un détail.
```suggestion
name: CI
```

AmauryCarrade commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,209 @@
+          path: dist
+      - name: prints alls files
+        run: |
+          ls -rtlh dist
```
Pourquoi ? Est-ce un test qui est resté ?

Si en fait c'est important et à garder, typo.

```suggestion
      - name: Print all dist files
        run: |
          ls -rtlh dist
```

firm1 committed:
fix coverage

AmauryCarrade commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,209 @@
+          coverage run --source='.' manage.py test -v=2 --keepdb --settings zds.settings.ci_test ${{ matrix.module }}
+
+      - name: Coveralls
+        uses: AndreMiras/coveralls-python-action@develop
```
Pourquoi dépendre d'une version de développement ? N'est-ce pas risqué ?

AmauryCarrade replied:
Rendu obsolète par un commit entre temps.

AmauryCarrade commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,209 @@
+      - name: Push documentation github
+        if: ${{ steps.extract_branch.outputs.branch == 'dev' }}
+        run : |
+          echo -e "machine github.com login $BOT_LOGIN\n password $BOT_PASSWORD" > ~/.netrc \
```
Utiliser des [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets) ici

Aussi, pourquoi ne pas utiliser uniquement `git config` pour configurer le mot de passe ?

Aussi, bis, il n'est plus nécessaire d'avoir un bot dédié pour mettre la doc dans la branche `gh-pages`, car GitHub Actions met à disposition un jeton permettant d'écrire sur toutes les branches. On peut donc faire quelque chose avec lui directement (et alors, ce sera `github-pages` qui fera le commit), ou [utiliser une action toute prête pour ça](https://github.com/zDevelopers/QuartzLib/blob/de127f8874ef935dcb3fec050d192d3b13d655c1/.github/workflows/docs.yml#L20-L26).

Aussi, ter, je me demandais initiaielement s'il n'était pas judicieux de mettre la construction de la doc à part (un peu comme sur le dernier lien), mais si on veut vérifier que la doc est correcte dans le cadre de tous les tests, ça fait sens de faire comme ça.

firm1 replied:
Oui j'ai vu passer les secrets au niveau de la doc, mais je n'ai pas les accès qui vont bien sur zds pour pouvoir les créer. Il faudrait que quelqu'un qui a les droits (qui puisse pinger pour ça ?) crée les variables qui vont bien et me communiques les infos.

AmauryCarrade replied:
~~Je ping `@zestedesavoir/pleins-pouvoirs` (si pas le temps, tu peux peut-être donner des droits temporaires ?)~~

Ensuite il suffit d'utiliser `${{ secrets.NOM_DU_SECRET }}`

AmauryCarrade replied:
Cela dit, au cas où tu as lu trop vite, ce n'est a priori plus nécessaire du tout vu que la branche `gh-pages` est dans le même dépôt.

> Aussi, bis, il n'est plus nécessaire d'avoir un bot dédié pour mettre la doc dans la branche `gh-pages`, car GitHub Actions met à disposition un jeton permettant d'écrire sur toutes les branches. On peut donc faire quelque chose avec lui directement (et alors, ce sera `github-actions` qui fera le commit), ou [utiliser une action toute prête pour ça](https://github.com/zDevelopers/QuartzLib/blob/de127f8874ef935dcb3fec050d192d3b13d655c1/.github/workflows/docs.yml#L20-L26).

AmauryCarrade replied:
Rendu obsolète par 6526a0bb1aadca18963b73c64b3f9b3fb0f2356b (un jour je lirai tout l'historique avant de répondre).

AmauryCarrade commented on scripts/install_zds.sh:
```diff
@@ -156,6 +148,8 @@ fi
 
 
 # virtualenv
+echo "===> $@"
```
Test oublié ou nécessité qui m'échappe ?

AmauryCarrade reviewed (changes requested):
Merci beaucoup pour tout ça ! J'ai quelques remarques et questions ci-dessous mais dans l'ensemble ça a l'air de plutôt bien tourner. Je laisse @Situphen compléter avec ses remarques, je n'ai fait qu'une revue rapide pour le moment.

D'un point de vue purement style, j'aurais quelques suggestions à faire en terme de formatage du fichier (en prenant plus d'espaces afin d'y voir plus clair) et du nom des différentes tâches (quand il n'en manque pas un tout court) afin de mieux s'y retrouver derrière quand on regarde le résultat des tests (certains détails et incohérences de nommages me titillent :D). Si tu veux, je pourrai directement faire un commit pour ça moi-même.

Aussi, j'avais, dans mes tests, [une version du script d'installation compatible avec les _folds_ de GitHub Actions](https://github.com/Situphen/zds-site/pull/10/checks?sha=b405501cf64cf91c75f55c0a652a3771268c477c), toujours dans l'objectif de rendre le log plus clair. Je pourrais le rapatrier (dans cette PR ou dans une autre, en soit c'est pas très urgent).

Enfin, à voir les durées, je me dis qu'on pourrait gagner sensiblement à diviser les tests de `tutorialv2` en deux parties équilibrées, mais ce serait plus pour une autre PR, je pense.

firm1 committed:
@Amaury remarks

firm1 commented:
Merci @AmauryCarrade pour ton retour rapide. J'ai poussé les plus évidentes (traces de debug qui étaient restée)

> j'aurais quelques suggestions à faire en terme de formatage du fichier (en prenant plus d'espaces afin d'y voir plus clair) et du nom des différentes tâches (quand il n'en manque pas un tout court) afin de mieux s'y retrouver derrière quand on regarde le résultat des tests (certains détails et incohérences de nommages me titillent :D). Si tu veux, je pourrai directement faire un commit pour ça moi-même.

Je veux bien un commit sur ces aspect.

> Aussi, j'avais, dans mes tests, une version du script d'installation compatible avec les folds de GitHub Actions, toujours dans l'objectif de rendre le log plus clair. Je pourrais le rapatrier (dans cette PR ou dans une autre, en soit c'est pas très urgent).

je ne suis pas sur de voir ce que tu qualifie de "fold github" si c'est l'hierarchisation des log, oui clairement. Si par contre, c'est la reprise du script sh qui fait un peu trop boite noire à mon gout, j'avoue ne pas être un grand fan.

> Enfin, à voir les durées, je me dis qu'on pourrait gagner sensiblement à diviser les tests de tutorialv2 en deux parties équilibrées, mais ce serait plus pour une autre PR, je pense.

Clairement oui.

AmauryCarrade commented:
> je ne suis pas sur de voir ce que tu qualifie de "fold github" si c'est l'hierarchisation des log, oui clairement. Si par contre, c'est la reprise du script sh qui fait un peu trop boite noire à mon gout, j'avoue ne pas être un grand fan.

En fait j'ai fait l'erreur d'écrire ça _avant_ la revue et avant de voir que tu n'utilises pas du tout le script d'installation ici. La remarque est donc rendue caduque.

Ça permettait, en utilisant le script effectivement, de réduire l'affichage par catégories afin de ne pas avoir un log de 4000 lignes, en masquant les logs dans des sections repliées par catégorie. Mais ce n'est pas trop applicable ici, sauf peut-être à certains endroits. Je relirai le _workflow_ et son log pour voir s'il y a des endroits où ce serait pertinent.

Concrètement c'est extrêmement simple à ajouter, c'est la même logique que le `set-output`. `::group::Nom du groupe` envoyé sur `stdout` démarre un groupe replié, et `::endgroup` l'arrête.

> Je veux bien un commit sur ces aspect.

Je ferai ça (mais pas maintenant, avec la réunion dans même pas un quart d'heure :p).

---

Autres points : il faudra qu'un propriétaire du dépôt rende obligatoire les bonnes tâches de CI, et rende facultative celle de Travis (et, au passage, déconnecte totalement ce dernier), dés que la PR aura été fusionnée.

Situphen commented:
> Autres points : il faudra qu'un propriétaire du dépôt rende obligatoire les bonnes tâches de CI, et rende facultative celle de Travis (et, au passage, déconnecte totalement ce dernier), dés que la PR aura été fusionnée.

Je m'en occuperai !

firm1 committed:
remplace le push de la doc

AmauryCarrade committed:
Améliore la forme du fichier de workflow

- Ajoute plein de lignes vide pour pouvoir plus facilement scanner le fichier.
- Renomme toute les sections pour mieux scanner le log de GitHub Actions.
- Explicite les versions utilisées dans ce log pour ne pas se faire surprendre.
- Ajoute des commentaires expliquant en une phrase ce que fait chaque job.

AmauryCarrade committed:
GitHub n'aime pas les commentaires, semble-t-il.

AmauryCarrade committed:
Revert "GitHub n'aime pas les commentaires […]"

This reverts commit 08b65c1265ec96f87ba2c48ab987e37dd8b251ed.

AmauryCarrade committed:
Petite erreur YAML passée inaperçue

AmauryCarrade committed:
Rassemble toutes les versions dans l'environnement.

- Toutes les versions des divers services utilisés par la CI
  sont précisées tout en haut du fichier, en tant que variables
  d'environnement.

AmauryCarrade committed:
Plus de cohérence de nommage avec GitHub

AmauryCarrade committed:
La tâche push_doc ne tourne pas du tout sur une PR

- Modifie la CI avec un test au niveau de la tâche directement, afin
  qu'elle ne soit pas du tout démarrée si on est sur une PR.

AmauryCarrade committed:
Petite erreur de syntaxe

AmauryCarrade committed:
Pas d'environnement dans les services

- GitHub Actions ne permet pas d'utiliser les variables d'environnement
  hors des noms, des tests (`if`) et des variables passées à des actions
  (`with`). Donc, pas dans les définitions de services. Par conséquent,
  nous sommes forcés de renseigner les versions en dur. Pour référence, j'ai
  gardé les versions en haut du fichier avec un commentaire précisant clairement
  qu'il faut mettre à jour les deux.

AmauryCarrade committed:
Final rename, I guess

AmauryCarrade committed:
Renomme test.yml en ci.yml

Le nom est plus représentatif de tout ce que ça fait (beaucoup plus que juste des tests).

AmauryCarrade commented:
Ok alors j'étais parti pour juste reformater un peu mais je suis allé un poiiil plus loin. J'ai tout séparé dans des commits donc en cas de désaccord il suffit de `git revert` !

- J'ai déplacé les références à des versions que j'ai pu vers les variables d'environnement. “Que j'ai pu”, car GitHub Actions n'autorise pas l'usage de ces variables hors des noms, des variables des actions (`with`), et des tests. Les versions d'ElasticSearch et de Memcached restent donc en dur, mais sont tout de même référencées en haut du fichier pour… référence, avec un commentaire précisant qu'il ne faut pas que les changer là.
- Plutôt que d'extraire la branche et d'exclure une étape de `push_doc` avec, j'ai transféré le test au niveau de `push_doc` lui-même, ce qui fait que la tâche ne démarre pas _du tout_ si on est sur une PR.  
  ![image](https://user-images.githubusercontent.com/1417570/99966767-14f76000-2d97-11eb-8528-34acf8f1f214.png)
- Dans la lignée d'un précédent renommage, j'ai renommé le fichier `tests.yml` en `ci.yml`, car c'est plus représentatif de ce qu'il fait (il est loin de ne faire que des tests).

AmauryCarrade commented:
Ayant plus participé que prévu je me retire des personnes devant approuver ou non (j'ai juste retiré ma précédente demande de changement désormais traitée, mais l'approbation n'en est que la conséquence, il faudrait autrui pour ça).

AmauryCarrade committed:
On n'est pas à un renomamge près

firm1 commented:
J'ai oublié de repasser ici préciser que tes modifications @AmauryCarrade sont les bienvenues. Si une ame charitable veut bien confirmer pour le merge de tout ça.

Situphen commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,286 @@
+
+      - name: Build and start zmarkdown
+        run: |
+          cd ./zmd && npm install pm2 && cd ..
```
Normalement cette ligne n'est pas nécessaire car `pm2` vient avec le paquet `zmarkdown` installé par `make zmd-install`

Situphen reviewed (commented):
Au niveau du code je n'ai rien à redire sauf ce petit détail de rien du tout ! :)

AmauryCarrade committed:
Retire une installation explicite inutile

Remarque de Situphen
